File: backend/src/main/java/com/gabaritaplus/api/entity/Question.java
package com.gabaritaplus.api.entity;

import com.gabaritaplus.api.entity.enums.DifficultyLevel;
import com.gabaritaplus.api.entity.enums.QuestionImportStatus;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "questions")
public class Question extends BaseEntity {

    @Column(nullable = false, length = 180)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String statement;

    @Column(name = "statement_html", columnDefinition = "TEXT")
    private String statementHtml;

    @Column(name = "image_url", length = 500)
    private String imageUrl;

    @Column(nullable = false, length = 80)
    private String subject;

    @Column(nullable = false, length = 120)
    private String topic;

    @Column(length = 120)
    private String subtopic;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private DifficultyLevel difficulty;

    @Column(nullable = false)
    private Integer year;

    @Column(nullable = false, length = 120)
    private String exam;

    @Column(length = 120)
    private String competency;

    @Column(length = 120)
    private String ability;

    @Column(columnDefinition = "TEXT")
    private String explanation;

    @Column(name = "correct_alternative", nullable = false, length = 1)
    private String correctAlternative;

    @Column(nullable = false, length = 80)
    private String source = "PLATFORM";

    @Column(name = "source_url", nullable = false, length = 1000)
    private String sourceUrl = "internal://legacy-question";

    @Column(name = "source_exam", nullable = false, length = 120)
    private String sourceExam;

    @Column(name = "source_year", nullable = false)
    private Integer sourceYear;

    @Column(name = "source_question_number")
    private Integer sourceQuestionNumber;

    @Column(name = "source_book_color", length = 40)
    private String sourceBookColor;

    @Column(name = "source_day")
    private Integer sourceDay;

    @Column(name = "source_page")
    private Integer sourcePage;

    @Column(name = "official_source_url", length = 1000)
    private String officialSourceUrl;

    @Column(name = "official_pdf_url", length = 1000)
    private String officialPdfUrl;

    @Column(name = "official_answer_key_url", length = 1000)
    private String officialAnswerKeyUrl;

    @Column(name = "official_page")
    private Integer officialPage;

    @Column(name = "validated_against_official_source", nullable = false)
    private Boolean validatedAgainstOfficialSource = false;

    @Column(name = "validated_at")
    private OffsetDateTime validatedAt;

    @Column(name = "external_provider", length = 120)
    private String externalProvider;

    @Column(name = "external_provider_url", length = 1000)
    private String externalProviderUrl;

    @Column(name = "external_question_id", length = 255)
    private String externalQuestionId;

    @Column(name = "external_license", length = 255)
    private String externalLicense;

    @Column(name = "imported_at")
    private OffsetDateTime importedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "import_batch_id")
    private ImportBatch importBatch;

    @Column(name = "statement_hash", nullable = false, length = 128)
    private String statementHash;

    @Enumerated(EnumType.STRING)
    @Column(name = "import_status", nullable = false, length = 30)
    private QuestionImportStatus importStatus = QuestionImportStatus.PUBLISHED;

    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Alternative> alternatives = new ArrayList<>();

    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<QuestionAsset> assets = new ArrayList<>();

    @PrePersist
    @PreUpdate
    private void applyImportDefaults() {
        if (sourceExam == null) {
            sourceExam = exam;
        }
        if (sourceYear == null) {
            sourceYear = year;
        }
        if (statementHash == null || statementHash.isBlank()) {
            statementHash = Integer.toHexString((title + "::" + statement).hashCode());
        }
        if (sourceUrl == null || sourceUrl.isBlank()) {
            sourceUrl = "internal://legacy-question";
        }
        if (source == null || source.isBlank()) {
            source = "PLATFORM";
        }
        if (importStatus == null) {
            importStatus = QuestionImportStatus.PUBLISHED;
        }
        if (validatedAgainstOfficialSource == null) {
            validatedAgainstOfficialSource = false;
        }
        if (importedAt == null) {
            importedAt = OffsetDateTime.now();
        }
    }
}

File: frontend/src/types/question.ts
import type { PageMetadata } from "@/types/api";

export type DifficultyLevel = "EASY" | "MEDIUM" | "HARD";
export type MasteryStatus = "NEW" | "LEARNING" | "REVIEW" | "MASTERED";
export type ReviewPriority = "HIGH" | "MEDIUM" | "LOW";
export type QuestionImportStatus =
  | "DRAFT"
  | "NEEDS_REVIEW"
  | "VALIDATED"
  | "AUTO_VALIDATED"
  | "PUBLISHED"
  | "INVALID";
export type AutoValidationStatus =
  | "SAFE_TO_AUTO_VALIDATE"
  | "NEEDS_HUMAN_REVIEW"
  | "AUTO_INVALID";
export type QuestionAssetType =
  | "IMAGE"
  | "GRAPH"
  | "TABLE"
  | "MAP"
  | "CHART"
  | "DIAGRAM"
  | "FORMULA"
  | "OTHER";

export interface QuestionAsset {
  id: number;
  questionId: number | null;
  alternativeId: number | null;
  type: QuestionAssetType;
  url: string | null;
  storagePath: string | null;
  originalFileName: string | null;
  sourcePage: number | null;
  cropX: number | null;
  cropY: number | null;
  cropWidth: number | null;
  cropHeight: number | null;
  altText: string | null;
  caption: string | null;
  checksum: string | null;
}

export interface Alternative {
  id: number;
  letter: string;
  text: string;
  html: string | null;
  assets: QuestionAsset[];
}

export interface Question {
  id: number;
  title: string;
  statement: string;
  statementHtml: string | null;
  imageUrl: string | null;
  subject: string;
  topic: string;
  subtopic: string | null;
  difficulty: DifficultyLevel;
  year: number;
  exam: string;
  competency: string | null;
  ability: string | null;
  source: string;
  sourceUrl: string;
  sourceExam: string;
  sourceYear: number;
  sourceQuestionNumber: number | null;
  sourceBookColor: string | null;
  sourceDay: number | null;
  sourcePage: number | null;
  officialSourceUrl: string | null;
  officialPdfUrl: string | null;
  officialAnswerKeyUrl: string | null;
  officialPage: number | null;
  validatedAgainstOfficialSource: boolean;
  validatedAt: string | null;
  externalProvider: string | null;
  externalProviderUrl: string | null;
  externalQuestionId: string | null;
  externalLicense: string | null;
  statementHash: string;
  importStatus: QuestionImportStatus;
  explanation: string | null;
  correctAlternative: string | null;
  favorite: boolean;
  answered: boolean | null;
  answeredCorrectly: boolean | null;
  assets: QuestionAsset[];
  alternatives: Alternative[];
  createdAt: string;
  updatedAt: string;
}

export interface ReviewQuestionSummary {
  id: number;
  title: string;
  source: string;
  sourceYear: number | null;
  sourceQuestionNumber: number | null;
  sourceBookColor: string | null;
  sourceDay: number | null;
  importStatus: QuestionImportStatus;
  validatedAgainstOfficialSource: boolean;
  externalProvider: string | null;
  importBatchId: number | null;
  subject: string;
  difficulty: DifficultyLevel;
  createdAt: string;
  importedAt: string | null;
  alternativesCount: number;
  assetsCount: number;
  autoValidationScore: number;
  autoValidationStatus: AutoValidationStatus;
  autoValidationWarnings: string | null;
  autoValidationErrors: string | null;
  brokenImageDetected: boolean;
  suspiciousTextDetected: boolean;
  requiresAssetReview: boolean;
}

export interface ReviewQuestionDetail extends Omit<Question, "favorite" | "answered" | "answeredCorrectly"> {
  importedAt: string | null;
  importBatchId: number | null;
  alternativesCount: number;
  assetsCount: number;
  autoValidationScore: number;
  autoValidationStatus: AutoValidationStatus;
  autoValidationErrors: string | null;
  autoValidationWarnings: string | null;
  autoValidatedAt: string | null;
  brokenImageDetected: boolean;
  suspiciousTextDetected: boolean;
  requiresAssetReview: boolean;
}

export interface ReviewQuestionFilters {
  page?: number;
  size?: number;
  sortBy?: string;
  direction?: "ASC" | "DESC";
  status?: QuestionImportStatus | "";
  source?: string;
  year?: number | "";
  subject?: string;
  autoValidationStatus?: AutoValidationStatus | "";
}

export interface ReviewQuestionListResponse {
  items: ReviewQuestionSummary[];
  metadata: PageMetadata;
}

export interface ReviewQuestionStatusPayload {
  importStatus: Exclude<QuestionImportStatus, "PUBLISHED">;
}

export interface ReviewOfficialValidationPayload {
  officialSourceUrl?: string;
  officialPdfUrl?: string;
  officialAnswerKeyUrl?: string;
  officialPage?: number | null;
}

export interface AutoValidationCounters {
  safe: number;
  needsReview: number;
  invalid: number;
  brokenImages: number;
  pendingInep: number;
}

export interface AutoValidationBatchResult {
  processed: number;
  safe: number;
  needsReview: number;
  invalid: number;
  published: number;
}

export interface OfficialExamSource {
  id: number;
  exam: string;
  year: number;
  day: number | null;
  bookColor: string | null;
  pdfUrl: string;
  answerKeyUrl: string | null;
  sourceUrl: string;
  localPdfPath: string | null;
  answerKeyMapJson: string | null;
  createdAt: string;
}

export interface OfficialExamSourcePayload {
  exam: string;
  year: number;
  day?: number | null;
  bookColor?: string | null;
  pdfUrl: string;
  answerKeyUrl?: string | null;
  sourceUrl: string;
  localPdfPath?: string | null;
  answerKeyMapJson?: string | null;
}

export interface OfficialValidationItem {
  questionId: number;
  title: string;
  sourceQuestionNumber: number | null;
  previousValidatedAgainstOfficialSource: boolean | null;
  newValidatedAgainstOfficialSource: boolean | null;
  previousScore: number | null;
  newScore: number | null;
  importStatus: QuestionImportStatus;
  autoValidationStatus: AutoValidationStatus;
  validatedAgainstOfficialSource: boolean;
  assetRecovered: boolean;
  recoveredAssets: number;
  newAutoValidationScore: number | null;
  newAutoValidationStatus: AutoValidationStatus | null;
  requiresAssetReview: boolean;
  brokenImageDetected: boolean;
  recoveryAttempted: boolean;
  officialSourceFound: boolean;
  pdfDownloaded: boolean;
  pdfSizeBytes: number | null;
  pdfPageCount: number | null;
  candidatePages: number[];
  selectedPage: number | null;
  pdfRendered: boolean;
  renderedWidth: number | null;
  renderedHeight: number | null;
  storageUploadAttempted: boolean;
  storageUploadSuccess: boolean;
  recoveryFailureReason: string | null;
  recoveryMethod: string | null;
  assetUrl: string | null;
  updated: boolean;
  warnings: string[];
  errors: string[];
}

export interface OfficialValidationReport {
  totalProcessed: number;
  processed: number;
  validated: number;
  skipped: number;
  failed: number;
  ambiguousOfficialSource: number;
  answerKeyMissing: number;
  answerKeyMismatch: number;
  updatedQuestions: number;
  needsReview: number;
  invalid: number;
  brokenImages: number;
  pendingAssets: number;
  pendingInep: number;
  assetRecovered: number;
  assetRecoveryFailed: number;
  items: OfficialValidationItem[];
}

export interface QuestionFilters {
  page?: number;
  size?: number;
  sortBy?: string;
  direction?: "ASC" | "DESC";
  search?: string;
  subject?: string;
  topic?: string;
  subtopic?: string;
  difficulty?: DifficultyLevel | "";
  year?: number | "";
  exam?: string;
  answered?: boolean | "";
  incorrectOnly?: boolean;
  favoritesOnly?: boolean;
}

export interface QuestionListResponse {
  items: Question[];
  metadata: PageMetadata;
}

export interface AnswerQuestionPayload {
  questionId: number;
  chosenAlternative: string;
  timeSpentSeconds: number;
}

export interface AnswerQuestionResult {
  id: number;
  questionId: number;
  chosenAlternative: string;
  correct: boolean;
  correctAlternative: string;
  explanation: string | null;
  attemptNumber: number;
  timeSpentSeconds: number;
  answeredAt: string;
}

export interface ErrorNotebookFilters {
  subject?: string;
  topic?: string;
  difficulty?: DifficultyLevel | "";
  masteryStatus?: MasteryStatus | "";
  priority?: ReviewPriority | "";
}

export interface ErrorNotebookEntry {
  id: number;
  questionId: number;
  questionTitle: string;
  subject: string;
  topic: string;
  difficulty: DifficultyLevel;
  errorCount: number;
  lastErrorAt: string | null;
  lastReviewedAt: string | null;
  nextReviewAt: string | null;
  masteryStatus: MasteryStatus;
  priority: ReviewPriority;
  updatedAt: string;
}

export interface UpdateErrorNotebookStatusPayload {
  masteryStatus: MasteryStatus;
}
